README.md
# PostCSS Easings

<img align="right" width="135" height="95"
     title="Philosopher’s stone, logo of PostCSS"
     src="https://postcss.org/logo-leftp.svg">

[PostCSS] plugin to replace easing name from [easings.net] to `cubic-bezier()`.

[easings.net]: http://easings.net/
[PostCSS]:     https://github.com/postcss/postcss

```css
.snake {
  transition: all 600ms ease-in-sine;
}
.camel {
  transition: all 600ms easeInSine;
}
```

```css
.snake {
  transition: all 600ms cubic-bezier(0.47, 0, 0.745, 0.715);
}
.camel {
  transition: all 600ms cubic-bezier(0.47, 0, 0.745, 0.715);
}
```

<a href="https://evilmartians.com/?utm_source=postcss-easings">
  <img src="https://evilmartians.com/badges/sponsored-by-evil-martians.svg"
       alt="Sponsored by Evil Martians" width="236" height="54">
</a>


## Usage

**Step 1:** Install plugin:

```sh
npm install --save-dev postcss postcss-easings
```

**Step 2:** Check your project for existed PostCSS config: `postcss.config.js`
in the project root, `"postcss"` section in `package.json`
or `postcss` in bundle config.

If you do not use PostCSS, add it according to [official docs]
and set this plugin in settings.

**Step 3:** Add the plugin to plugins list:

```diff
module.exports = {
  plugins: [
+   require('postcss-easings'),
    require('autoprefixer')
  ]
}
```

[official docs]: https://github.com/postcss/postcss#usage

Also, you can get all built-in easings:

```js
require('postcss-easings').easings;
```


## Options

### `easings`

Allow to set custom easings:

```js
    require('postcss-easings')({
      easings: { easeJump: 'cubic-bezier(.86,0,.69,1.57)' }
    })
```

The plugin will convert custom easing name between camelCase and snake-case.
So the example below adds `easeJump` and `ease-jump` easings.

Custom easing names must start from `ease` and contain only letters and `-`.

You can create custom easing on [cubic-bezier.com].

[cubic-bezier.com]: http://cubic-bezier.com/


CHANGELOG.md
# Change Log
This project adheres to [Semantic Versioning](http://semver.org/).

## 4.0.0
* Removed Node.js 14 support.
* Updated easings from easings.net (by @sidewayss).
* Fixed custom easings with numbers support (by @sidewayss).

## 3.0.1
* Added funding links.

## 3.0
* Moved to PostCSS 8.
* Moved `postcss` to `peerDependencies`.

## 2.0
* Use PostCSS 7.
* Remove Node.js 4 support.

## 1.0.1
* Do not replace values in `url()` (by Fangzhou Li).

## 1.0
* Use PostCSS 6.0.

## 0.3
* Use PostCSS 5.0.
* Fix error reports.

## 0.2
* Support PostCSS 4.1 API.

## 0.1
* Initial release.


test/test.js
var postcss = require('postcss');
var expect  = require('chai').expect;

var easings = require('../');

var test = function (input, output, opts) {
    expect(postcss(easings(opts)).process(input).css).to.eql(output);
};

describe('postcss-easings', function () {

    it('replaces easings by camel case name', function () {
        test('a { transition: all 1s easeInSine }',
             'a { transition: all 1s cubic-bezier(0.47, 0, 0.745, 0.715) }');
    });

    it('replaces easings by snake case name', function () {
        test('a { transition: all 1s ease-in-sine }',
             'a { transition: all 1s cubic-bezier(0.47, 0, 0.745, 0.715) }');
    });

    it('replaces multiple easings in out value', function () {
        test('a { transition: ease-in-sine, easeInOutExpo }',
             'a { transition: cubic-bezier(0.47, 0, 0.745, 0.715), ' +
                             'cubic-bezier(1, 0, 0, 1) }');
    });

    it('allows to add custom easings', function () {
        test('a { transition: ease-my, easeMy }',
             'a { transition: 1, 1 }',
             { easings: { easeMy: '1' } });
    });

    it('allows to add custom easings with snake name', function () {
        test('a { transition: ease-my, easeMy }',
             'a { transition: 1, 1 }',
             { easings: { 'ease-my': '1' } });
    });

    it('checks custom easings name', function () {
        expect(function () {
            easings({ easings: { my: '1' } });
        }).to.throw(/^Custom easing my has bad name/);
    });

    it('exports easings', function () {
        expect(easings.easings.easeInSine)
            .to.eql('cubic-bezier(0.47, 0, 0.745, 0.715)');
    });

});


index.test.js
let { equal, throws } = require('uvu/assert')
let postcss = require('postcss')
let { test } = require('uvu')

let plugin = require('./')

function run(input, output, opts) {
  let result = postcss([plugin(opts)]).process(input, { from: '/test.css' })
  equal(result.css, output)
  equal(result.warnings().length, 0)
}

test('replaces easings by camel case name', async () => {
  await run(
    'a { transition: all 1s easeInSine }',
    'a { transition: all 1s cubic-bezier(0.12, 0, 0.39, 0) }'
  )
})

test('replaces easings in custom properties', async () => {
  await run(
    ':root { --animation: easeInSine }',
    ':root { --animation: cubic-bezier(0.12, 0, 0.39, 0) }'
  )
})

test('parses regular functions', async () => {
  await run(
    'a { transition: all 1s cubic-bezier(0.47, 0, 0.745, 0.715) }',
    'a { transition: all 1s cubic-bezier(0.47, 0, 0.745, 0.715) }'
  )
})

test('ignores unknown names', async () => {
  await run(
    'a { transition: all 1s easeInSine1 }',
    'a { transition: all 1s easeInSine1 }'
  )
})

test('replaces easings by snake case name', async () => {
  await run(
    'a { transition: all 1s ease-in-sine }',
    'a { transition: all 1s cubic-bezier(0.12, 0, 0.39, 0) }'
  )
})

test('replaces multiple easings in out value', async () => {
  await run(
    'a { transition: ease-in-sine, easeInOutExpo }',
    'a { transition: cubic-bezier(0.12, 0, 0.39, 0), ' +
      'cubic-bezier(0.87, 0, 0.13, 1) }'
  )
})

test('allows to add custom easings', async () => {
  await run('a { transition: ease-my, easeMy }', 'a { transition: 1, 1 }', {
    easings: { easeMy: '1' }
  })
})

test('allows to add custom easings with snake name', async () => {
  await run('a { transition: ease-my, easeMy }', 'a { transition: 1, 1 }', {
    easings: { 'ease-my': '1' }
  })
})

test('allows to add custom easings without separation', async () => {
  await run('a { transition: easemy }', 'a { transition: 1 }', {
    easings: { easemy: '1' }
  })
})

test('checks custom easings name', () => {
  throws(() => {
    plugin({ easings: { my: '1' } })
  }, /^Custom easing my has bad name/)
})

test('checks another custom easings name with whitespace', () => {
  throws(() => {
    plugin({ easings: { 'ease- In': '1' } })
  }, /^Custom easing ease- In has bad name/)
})

test('add custom easings with numbers in name', async () => {
  await run('a { transition: ease4-you, ease4You }', 'a { transition: 1, 1 }', {
    easings: { ease4You: '1' }
  })
})

test('add custom easings with numbers in snake name', async () => {
  await run('a { transition: ease-4-you, ease4You }', 'a { transition: 1, 1 }', {
    easings: { 'ease-4-you': '1' }
  })
})

test('exports easings', () => {
  equal(plugin.easings.easeInSine, 'cubic-bezier(0.12, 0, 0.39, 0)')
})

test.run()


index.js
let parser = require('postcss-value-parser')

const EASINGS = {
  easeInBack: 'cubic-bezier(0.36, 0, 0.66, -0.56)',
  easeInCirc: 'cubic-bezier(0.55, 0, 1, 0.45)',
  easeInCubic: 'cubic-bezier(0.32, 0, 0.67, 0)',
  easeInExpo: 'cubic-bezier(0.7, 0, 0.84, 0)',
  easeInOutBack: 'cubic-bezier(0.68, -0.6, 0.32, 1.6)',
  easeInOutCirc: 'cubic-bezier(0.85, 0, 0.15, 1)',
  easeInOutCubic: 'cubic-bezier(0.65, 0, 0.35, 1)',
  easeInOutExpo: 'cubic-bezier(0.87, 0, 0.13, 1)',
  easeInOutQuad: 'cubic-bezier(0.45, 0, 0.55, 1)',
  easeInOutQuart: 'cubic-bezier(0.76, 0, 0.24, 1)',
  easeInOutQuint: 'cubic-bezier(0.83, 0, 0.17, 1)',
  easeInOutSine: 'cubic-bezier(0.37, 0, 0.63, 1)',
  easeInQuad: 'cubic-bezier(0.11, 0, 0.5, 0)',
  easeInQuart: 'cubic-bezier(0.5, 0, 0.75, 0)',
  easeInQuint: 'cubic-bezier(0.64, 0, 0.78, 0)',
  easeInSine: 'cubic-bezier(0.12, 0, 0.39, 0)',
  easeOutBack: 'cubic-bezier(0.34, 1.56, 0.64, 1)',
  easeOutCirc: 'cubic-bezier(0, 0.55, 0.45, 1)',
  easeOutCubic: 'cubic-bezier(0.33, 1, 0.68, 1)',
  easeOutExpo: 'cubic-bezier(0.16, 1, 0.3, 1)',
  easeOutQuad: 'cubic-bezier(0.5, 1, 0.89, 1)',
  easeOutQuart: 'cubic-bezier(0.25, 1, 0.5, 1)',
  easeOutQuint: 'cubic-bezier(0.22, 1, 0.36, 1)',
  easeOutSine: 'cubic-bezier(0.61, 1, 0.88, 1)'
}

function toSnake (str) {
  return str.replace(/[A-Z]/g, letter => {
    return '-' + letter.toLowerCase()
  })
}

function toCamel (str) {
  return str.replace(/-(.)/g, letter => {
    return letter[1].toUpperCase()
  })
}

for (let camel of Object.keys(EASINGS)) {
  EASINGS[toSnake(camel)] = EASINGS[camel]
}

const EASING_NAME = /^ease([\w-]+)$/

module.exports = (opts = {}) => {
  let locals = {}
  if (opts.easings) {
    for (let name in opts.easings) {
      if (!EASING_NAME.test(name)) {
        throw new Error(
          `Custom easing ${name} has bad name. ` +
            'Name should start with `ease` and contain ' +
            'only letters, numbers, underscore and dashes'
        )
      }
      locals[name] = opts.easings[name]
      if (name.includes('-')) {
        locals[toCamel(name)] = opts.easings[name]
      } else if (/[A-Z]/.test(name)) {
        locals[toSnake(name)] = opts.easings[name]
      }
    }
  }

  return {
    Declaration (decl) {
      if (!decl.value.includes('ease')) return
      let root = parser(decl.value)
      let changed = false
      root.nodes = root.nodes.map(node => {
        let value = node.value
        if (node.type === 'word' && EASING_NAME.test(value)) {
          if (locals[value] || EASINGS[value]) {
            changed = true
            node.value = locals[value] || EASINGS[value]
          }
        }
        return node
      })
      if (changed) {
        decl.value = root.toString()
      }
    },
    postcssPlugin: 'postcss-easings'
  }
}
module.exports.postcss = true

module.exports.easings = EASINGS
